sympy/polys/rings.py

```diff
@@ -1100,61 +1021,55 @@ def __rsub__(p1, n):
         >>> from sympy.polys.rings import ring
 
         >>> _, x, y = ring('x, y', ZZ)
-        >>> p = x + y
-        >>> 4 - p
-        -x - y + 4
-
+        >>> p = x + y**2
+        >>> p**3
+        x**3 + 3*x**2*y**2 + 3*x*y**4 + y**6
         """
-        ring = p1.ring
-        try:
-            n = ring.domain_new(n)
-        except CoercionFailed:
-            return NotImplemented
-        else:
-            p = ring.zero
-            for expv in p1:
-                p[expv] = -p1[expv]
-            p += n
-            # p._check()
-            return p
+        if not isinstance(n, int):
+            raise TypeError("exponent must be an integer, got %s" % n)
+        elif n < 0:
+            raise ValueError("exponent must be a non-negative integer, got %s" % n)
 
-    def __mul__(p1, p2):
-        """Multiply two polynomials.
+        ring = self.ring
+
+        if not n:
+            if self:
+                return ring.one
+            else:
+                raise ValueError("0**0")
+        elif len(self) == 1:
+            return self._pow_single_term(n)
+
+        return self._pow_int(n)
```

Comment:
I don't think that the code before `_pow_int` here makes sense. The subclass can implement these things. The FLINT implementation here would just be:
```python
def _pow_int(self, n):
    return self._from_rep(self._rep ** n)
```
There would be no reason to have a separate `_pow_single_term` method.

Reply:
Of course the Python subclass can have a `_pow_single_term` method although I'm not sure what the advantage would be of having a method rather than it just being something that is handled inline in `_pow_int`.